src/deviation.jl

```diff
@@ -43,46 +43,26 @@ end
 """
     sqL2dist(a, b)
 
-Compute the squared L2 distance between two arrays: ``\\sum_{i=1}^n |a_i - b_i|^2``.
-Efficient equivalent of `sumabs2(a - b)`.
+Compute the squared L2 distance between `a` and `b`: ``\\sum_{i=1}^n |a_i - b_i|^2``.
 """
-function sqL2dist(a::AbstractArray{T}, b::AbstractArray{T}) where T<:Number
-    n = length(a)
-    length(b) == n || throw(DimensionMismatch("Input dimension mismatch"))
-    r = 0.0
-    for i = 1:n
-        @inbounds r += abs2(a[i] - b[i])
-    end
-    return r
-end
-
+sqL2dist(a, b) = sum(abs2.(a .- b))
```

Comment:
Better do `sum(abs2(x - y) for (x, y) in zip(a, b)))`. Same elsewhere.

Reply:
I  benchmarked a number of options around this.
That one is slower.
Constructing a `zip` generally needs to be avoided to get best performance


```
julia> @btime sum(abs2(x-y) for (x,y) in zip($a,$b));
  796.989 ns (3 allocations: 64 bytes)
```

vs the current code I have (copy pasted from above)

```
julia> @btime sum(abs2.($a.-$b))
  445.434 ns (1 allocation: 7.94 KiB)
16044
```

Reply:
Hmm, I don't get the same result here:
```julia
julia> @btime sum(abs2(x-y) for (x,y) in zip($a,$b));
  133.841 μs (3 allocations: 64 bytes)

julia> @btime sum(abs2.($a.-$b));
  158.430 μs (2 allocations: 781.33 KiB)
```

Also, `.-` only works for arrays, so that defeats the point of the PR (even if we could use it in a specialized method if it was faster).

Reply:
I am testing on julia 1.1 that might be the source of differences.
(Or it could be system dependent, I feel like cache might be involved)

It was my understanding that as of julia 1.0 broadcast defaults to iterating.
Thus this is general, no?

```
julia> Iterators.drop([1,2,3], 1) .- [10,20]
2-element Array{Int64,1}:
  -8
 -17
```

Reply:
Funny. The second variant is slower here both with 1.1 and master. I have a Skylake CPU. Do you get the same result if you wrap the code in functions?

> It was my understanding that as of julia 1.0 broadcast defaults to iterating.
> Thus this is general, no?

No, that's only the case for `HasShape` iterators:
```julia
julia> (i for i in [1,2,3]) .- [10,20]
ERROR: DimensionMismatch("arrays could not be broadcast to a common size")
```

Anyway I think we should use the generator approach as it avoids an intermediate allocation. Then it would be interesting to file an issue in Julia to investigate why it's faster on your machine.

Reply:
Avoiding the intermidate allocation may not be worth it if it hurt cache locality.

II also have a Skylake CPU.
```
julia> versioninfo()
Julia Version 1.1.0
Commit 80516ca202 (2019-01-21 21:24 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.1 (ORCJIT, skylake)
```

Reply:
Looking at the code, what happens is that the second version is able to use SIMD, while the first one isn't. I guess depending on the CPU the gain due to SIMD can offset the overhead of making a copy, or not. It's hard to decide which approach should be the fastest in general without testing more CPUs. But in theory I imagine LLVM could improve and be able to use SIMD without making a copy with the first version.

Reply:
Can you also try the original implementation after adding `@inbounds @simd` in front of the loop? In my tests, it's even faster. Using a single index for both arrays allows adding `@simd`, which isn't possible with `zip` (we could use `eachindex`, though, which would be more general).

Then we could keep both a specialized method for arrays and a more general one.

Reply:
Predicting the future is hard.
 - LLVM could improve to use SIMD in the first version
 - Julia could improve to avoid the copy in the second one
 - Computers with better SIMD could dominate the market as time moves forward.

Reply:
> Can you also try the original implementation after adding @inbounds @simd in front of the loop? 

At some point maybe. Tonight no.

more broadly I wonder if we should just deprecate all these into Distances.jl ?

Reply:
Yes, prediction is hard. That's why I'd favor a simpler criterion: avoid allocating a copy. If you pass a very large vector, you don't necessarily have the space to allocate a copy (it could also be mmapped), and computing a simple sum shouldn't require that. And indeed the docs mention this as a rule: http://juliastats.github.io/StatsBase.jl/latest/deviation.html#Computing-Deviations-1 (at the bottom)

But I agree it could make sense to deprecate all of these. Let's hear what others think.

Reply:
Any news on this?

Reply:
I was convinced by your arguments to deprecate them here.
And then blocked by my desire to create a better way to do deprecations.

Reply:
What better way? :-/

Reply:
I have this plan for a package
https://github.com/invenia/Deprecations.jl/pull/1/files

Reply:
Hmm, I'm not sure we want to take a new dependency just for deprecations. The current mechanism sounds good enough to me.

Reply:
Bump.

Reply:
I don't think I want to deprecate things in this PR.
That should be a different PR

Reply:
But changing the implementation of something we want to deprecate immediately sounds like unnecessary churn. Does this PR touch functions that shouldn't be deprecated?

Reply:
yes, `psnr` at least

Reply:
OK, as you like. But you're going to spend much more time changing functions that are to be deprecated than on  `psnr`.